rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,258 @@
+
+## Details
+
+### When is a Migration required?
+
+Migrations are not needed for every change to a backend resource. Adding a new field to the resource should be a backward compatible operation by nature of Protocol Buffers and does not require an actual migration. Any major changes to a resource that alter its shape, change its encoding, or moving a single resource into several other resources are candidates for a migration.
+
+### Persistent Migrations
+
+All new migrations MUST exist in perpetuity to allow for the correct migrations to be applied regardless of which version of Teleport is being upgraded from and to. Migrations MUST be numbered and applied in sequence. The order of migrations MUST not change and a migration MUST not be altered once it has been included in a release.
+
+To make tracking and discovering migrations easier all migrations MUST be placed in `lib/auth/migrations`. Each migration should be named `$number-migration-description.go` and contain a single migration function, see the Migrations section below for details. The following is an example of four migrations:
+
+```
+./lib/auth/migrations/
+├── 0001-initial_migration.go
+├── 0002-another_migration.go
+├── 0003-some_other_migration.go
+└── 0004-yet_another_migration.go
+```
+
+### Migration History
+
+Migration history is to be stored in the backend to enable tracking which migrations have been applied, when, and whether they were successful or not.
+The key `/migrations/<migration_number>` will store the history of a migration.
+
+Cluster admins may view the migration history via `tctl migrations ls`.
+
+### Migrating Keys
+
+All migrations MUST not be performed in place, instead migrations MUST migrate both the keys and values in the backend.
+For example, to migrate the data stored in key `some/path/to/resource/<ID>` we must leave the original value at `some/path/to/resource/<ID>` as is and write the migrated value to `some/new/path/to/resource/<ID>`.
+This allows older versions of Teleport to still operate on `some/path/to/resource/<ID>`, while newer versions can first attempt reads from `some/new/path/to/resource/<ID>`, and fall back to reading from `some/path/to/resource/<ID>` if the migrated key does not exist yet.
+
+Migrations MUST also not delete the key being migrated for at least one major version of Teleport. This allows users to upgrade to and downgrade from a single major version, but prevents any downgrading beyond that.
+Continuing from the example above, if `some/path/to/resource/<ID>` is migrated to `some/new/path/to/resource/<ID>` in v1.0.0 a follow up migration should be added in v2.0.0 to delete the keys under `some/path/to/resource` that were migrated in v1.0.0.
```

Comment:
Alternatively to custom paths, we can consider AWS global secondary index that in many ways act like materialized views including projecting fields:

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/GSI.html

You can store an old version of the value in the attribute, and let old auth server once it switched to read from an old index. While new auth servers will read from the source.

Reply:
@rosstimothy we can narrow the scope if we want. We only need to support zero downtime migrations for DynamoDB, as that's what our cloud is using, if that simplifies design and makes Teleport cloud more competitive.